proof-of-liquidity/src/main/java/icon/inflation/score/pol/ProofOfLiquidity.java

```diff
@@ -0,0 +1,118 @@
+    public static final VarDB<Address> balancedRewards = Context.newVarDB("BALANCED_REWARDS", Address.class);
+
+    private static BigInteger TOTAL_SHARE = EXA;
+    private static String NAME = "ICON Proof of liquidity Manager";
+
+    private static boolean distributing = false;
+
+    public ProofOfLiquidity(Address staking, Address sICX, Address balancedRewards) {
+        ProofOfLiquidity.staking.set(staking);
+        ProofOfLiquidity.sICX.set(sICX);
+        ProofOfLiquidity.balancedRewards.set(balancedRewards);
+    }
+
+    @External(readonly = true)
+    public String name() {
+        return NAME;
+    }
+
+    @External(readonly = true)
+    public LiquidityDistribution[] getDistributions() {
+        int size = distribution.size();
+        LiquidityDistribution[] _distribution = new LiquidityDistribution[size];
+        for (int i = 0; i < size; i++) {
+            _distribution[i] = distribution.get(i);
+        }
+
+        return _distribution;
+    }
+
+    @External
+    public void configureDistributions(LiquidityDistribution[] _distribution) {
+        onlyOwner();
+        BigInteger sum = BigInteger.ZERO;
+        DBUtils.clear(distribution);
+
+        for (LiquidityDistribution dist : _distribution) {
+            Context.require(dist.share.compareTo(BigInteger.ZERO) > 0, Errors.NEGATIVE_PERCENTAGE);
+            sum = sum.add(dist.share);
+            distribution.add(dist);
+        }
+
+        Context.require(sum.equals(TOTAL_SHARE), Errors.INVALID_SUM);
+    }
+
+    @External
+    public void distribute() {
+        Context.require(!distributing, Errors.NO_REENTRY);
+
+        BigInteger balance = Context.getBalance(Context.getAddress());
+        Context.require(balance.compareTo(BigInteger.ZERO) > 0, Errors.EMPTY_BALANCE);
+
+        int size = distribution.size();
+        Context.require(size > 0, Errors.NOT_CONFIGURED);
+
+        distributing = true;
+
+        Context.call(balance, staking.get(), "stakeICX", null, null);
+        BigInteger amount = Context.call(BigInteger.class, sICX.get(), "balanceOf", Context.getAddress());
+        BigInteger sum = BigInteger.ZERO;
+
+        JsonArray data = new JsonArray();
+        for (int i = 0; i < size; i++) {
+            LiquidityDistribution dist = distribution.get(i);
+            // Dev note: This way of calculating will create remaining dust. But since we
+            // are dealing with large amounts this should not matter and a good trade of for
+            // simplicity
+            BigInteger share = dist.share.multiply(amount).divide(TOTAL_SHARE);
+            sum = sum.add(share);
+            JsonValue jsonData = new JsonObject()
+                .add("source", dist.source)
+                .add("amount", share.toString());
+            data.add(jsonData);
+        }
+
+        Context.call(sICX.get(), "transfer", balancedRewards.get(), sum, data.toString().getBytes());
+        distributing = false;
```

Comment:
Maybe we can consider implementing an event log after the distribution process. This addition would facilitate filtering the logs specifically for the distribution of this particular contract.

Reply:
Yeah that might come in handy in the future

Reply:
On second thoughts we already have many logs that can be used for this and would be redundant. sICX and ICX transfers do raise event logs, that show all data and destinations.